models/mace/json_to_extxyz.py

```diff
@@ -1,47 +1,46 @@
-"""This script converts the MPTrj relaxation trajectories downloaded from
-https://figshare.com/articles/dataset/23713842 from JSON to extended XYZ format.
-"""
-
+import copy
 import json
 import os
+import urllib.request
 
+import ase.io
+import ase.units
 import numpy as np
-from ase.io import read, write
 from pymatgen.core import Structure
 from tqdm import tqdm
 
-from matbench_discovery import FIGSHARE
-
 __author__ = "Yuan Chiang"
 __date__ = "2023-08-10"
 
 module_dir = os.path.dirname(__file__)
-in_dir = f"{FIGSHARE}/mptrj_2022.9_full.json"
-out_dir = f"{module_dir}/mptrj-2022.9"
-os.makedirs(out_dir, exist_ok=True)
-combined = []
+mptrj_path = "./MPtrj_2022.9_full.json"
+urllib.request.urlretrieve(
+    "https://figshare.com/ndownloader/files/41619375", mptrj_path
+)
 
+with open(mptrj_path) as file:
+    json_data = json.load(file)
 
-with open(in_dir) as json_file:
-    json_data = json.load(json_file)
+out_dir = f"{module_dir}/mptrj-gga-ggapu"
+os.makedirs(out_dir, exist_ok=True)
+combined = []
 
 for material_id in tqdm(json_data):
-    for trajectory_id in json_data[material_id]:
-        out_xyz = f"{out_dir}/{material_id}.extxyz"
+    xyz_path = f"{out_dir}/{material_id}.extxyz"
 
-        if os.path.isfile(out_xyz):
-            traj = read(out_xyz, index=":", format="extxyz")
-            combined.append(traj)
-            continue
+    if os.path.isfile(xyz_path):  # read already converted file
+        traj = ase.io.read(xyz_path, index=":", format="extxyz")
+        combined.extend(traj)
+        continue
 
-        block = json_data[material_id][trajectory_id]
+    for trajectory_id in json_data[material_id]:
+        block = copy.deepcopy(json_data[material_id][trajectory_id])
```

Comment:
Is `copy.deepcopy` really needed? This might slow down the conversion substantially. Is it to avoid errors on processing twice? Maybe better to keep a `dict` that maps processed IDs to processed results. That way you can check in constant time whether you should skip an ID.

Reply:
since I use dict.pop the error will raise if size of dictionary changes within a for loop. copy will avoid that. The speed is doable for now it is roughly one hour processing time

Reply:
Always good to leave why comments in the code. Why are you doing something (if it's not obvious).

2nd question: Why are you using `pop`? What's wrong with `get`?

Reply:
we don’t want repeated forces, stress, magmoms, energy show up in atoms.info

Reply:
I don't follow. `get` also returns `None`?

Reply:
Yes, but `get` will not remove the key value pair from the dictionary and those will all go into atoms.info

https://github.com/janosh/matbench-discovery/blob/af1d7ad6fb9a4652fe8d5f9dd70d0117c424d959/models/mace/json_to_extxyz.py#L60-L61

Reply:
I see. Let's just use `set` arithmetic to exempt those keys from double processing then.